docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,157 @@
+}
+```
+**ObjectMeta.ParentReferences**: links the resource to the parent resources. For example, a replica set `R` created by a deployment `D` should have an entry in ObjectMeta.ParentReferences pointing to `D`. The link should be set when the child object is created. It can be updated after the creation.
+
+```
+type Tombstone struct {
+	UID types.UID
+}
+```
+**Tombstone**: a tombstone is created when an object is deleted and the user requires the children to be orphaned.
+**Tombstone.UID**: the UID of the original object.  
+
+## New components
+The only new component is the Garbage Collector, which consists of a scanner, a garbage processor, and a propagator.
+* Scanner:
+  * Uses the discovery API to detect all the resources supported by the system.
+  * Periodically scans all resources in the system and adds each object to the *Dirty Queue*.
+
+* Garbage Processor:
+  * Consists of the *Dirty Queue* and workers.
+  * Each worker:
+    * Dequeues an item from *Dirty Queue*
+    * Checks if the item’s parents still exist. 
+    * Checks for a tombstone of parents that don’t exist
+    * If there is no parent nor tombstone, requests the API server to delete the item.
+    * If tombstone does exist, removes the parent from the object’s ParentReferences.
+
+* Propagator:
+  * The Propagator is for optimization, not for correctness.
+  * Maintains a DAG of parent-child relations. This DAG stores only name/uid/orphan triplets, not the entire body of every item.
+  * Consists of an *Event Queue* and a single worker.
+  * Watches for create/update/delete events for all resources, enqueues the events to the *Event Queue*.
+  * Worker:
+    * Dequeues an item from the *Event Queue*.
+    * If the item is an creation or update, then updates the DAG accordingly.
+      * If the object has a parent and the parent doesn’t exist in the DAG yet, then apart from adding the object to the DAG, also enqueues the object to the *Dirty Queue*.
+    * If the item is a deletion, then removes the object from the DAG, and enqueues all its children to the *Dirty Queue*.
+  * The propagator shouldn't need to do any RPCs, so a single worker should be sufficient. This makes locking easier.
+  * With the Propagator, we *only* need to run the Scanner when starting the Propagator to populate the DAG and the *Dirty Queue*.
+
+## Changes to existing components
+* Storage: we should add a REST storage for Tombstones. The index should be UID rather than namespace/name.
+
+* API Server: when handling a deletion request, if DeleteOptions.Orphaning is true, then the API Server either creates a tombstone with TTL if the tombstone doesn't exist yet, or updates the TTL of the existing tombstone. The API Server deletes the object after the tombstone is created.
+
+* Controllers: when creating child objects, controllers need to fill up the their ObjectMeta.ParentReferences field. Objects that don’t have a parent should have the namespace object as the parent.
```

Comment:
Why is `ObjectMeta.Namespace/ObjectMeta.NamespaceUID` not sufficient?  Since we were allowing ParentReferences to change post creation, but we do not allow namespace parent relationship to change, this seems error prone...